### ISSUE TITLE
[Meta]: Screen Reader problems

### ISSUE THREAD

**Issue by TheTechnician27** [PCSX2/pcsx2#13303]
### Description

This is an ongoing meta issue for problems with screen readers in PCSX2. Upstream issues (e.g. with Qt) can also be noted here but should be distinguished where possible. I'm testing on the popular screen reader [Orca](https://en.wikipedia.org/wiki/Orca_(assistive_technology)), but other popular screen readers ought to be tested too.

### Reason

PCSX2 anecdotally – and as a statistical corollary from the way those with disabilities don't live sequestered in a void where fun doesn't exist – has a sizable base of visually impaired users. However, it seems like PCSX2's various and often severe problems with screen reader compatibility haven't been well-documented. While many of these are isolated hiccups, some of them fundamentally break the ability of the user to properly navigate the emulator. Some of these are even intersectional, like the tabbing issues that also affect those with motor impairments.

I suspect the ostensible lack of major effort is attributable to a combination of 1) screen reader software being completely unfamiliar to most project members (myself included until this investigation) and 2) difficulty conceptualizing enjoying a medium so fundamentally visual with a visual impairment. Ideally, however, once properly set up by the user, PCSX2 is an even more accessible way to experience the PS2 for this audience; it's one giant accessibility feature. We should do our best to foster that for the users we already have and to open the gates to new ones who feel locked out.

"If you build it, [they] will come."

## Settings

- [ ] Settings pages don't start out with any of the items focused (instead the two buttons at the bottom are), and you can't tab out of the sidebar selection, so you first need to click around in the window to find a setting to "anchor" to.
- [ ] Tabbing through many of the settings windows doesn't loop, instead terminating at 'Close'. However, some *do* loop.
- [x] (fixed via #13388) The ordering of settings in the `.ui` source files means that tabbing order is basically non-existent. For example, open `Emulation`, "anchor" yourself at Normal Speed by clicking it, and then keep tabbing and see the garbage, nonsense path it takes. (easy fix and obviously important not just for visual impairments but also motor impairments; currently working on it in a PR)
- [ ] Related to focusing, hovering over e.g. the checkbox options neither reads the option name nor reads any of the help dialog at the bottom.
- [ ] It's impossible to read out the help dialog at the bottom, because it's only there when moused over the option, which doesn't read that dialog.
- [x] (fixed via #13388) For issues like `Graphics` > `Vertical Stretch`, it's impossible to read what these options do. It just tells me it's a Spin Box with 100%, but it doesn't tell me what the spin box is doing. Again related to (5). This is the case for dropdown lists too like "Blending" and even the case for extremely basic options like "Internal Resolution". I tested the RTC spinbox that happens to work, but I think that's only because it's in a group titled "Real-Time Clock".
- [ ] The Game List Background trash bin button doesn't respond with any sort of feedback when clicked.
- [ ] The way that we choose to show the timeout text from within the controller and hotkey mapping buttons means the screen reader dutifully refreshes every second and helpfully tells the user: "push button slash push button slash push button slash push button slash push button slash".
- [ ] When clicking the `Verify` button for a game, the text "Computing hash for track 1..." is not read out.
- [ ] Clicking `Automatic Mapping` in the controller settings doesn't announce what menu the user is on (hovering doesn't announce it either).
- [ ] Clicking buttons like `Bindings`, `Macros`, and `Settings` from within controller settings doesn't tell me the menu I'm actually on (hover doesn't work).
- [ ] In `Mapping Settings` in controllers, it's hard to get `Ignore Inversion` into focus if you don't already know 1) *that* it's there, 2) where it is and 3) where your mouse cursor is relative to it.
- [ ] In ternary checkboxes in per-game settings, a checked box returns "Checked", but an unchecked box returns "Partially checked" while the grey box returns "Unchecked".

Opposite of an issue: tabbing over a button in the controller mapping gives great information. This seems like a really good example to look at when considering current issues.

## Main Window

- [ ] The toolbar buttons do not read when hovered over or clicked.
- [ ] The filter buttons near the search bar don't respond on hover, nor do they tell you the name of the box when you open them.
- [ ] The Game List correctly says the number of items (albeit doesn't tell you the headers), but the Game Grid only announces the number of games currently rendered on-screen. So for example, Game List might report 108 games while Game Grid reports only 27.
- [ ] The nametag button to toggle game names in the game grid doesn't respond to hover or click, and this is especially important because disabling names stops the screen reader from reading those titles on focus.
- [ ] Text in e.g. `Help` > `About` cannot be highlighted.
- [ ] Because the `About Qt` window has an `OK` button rather than a `Close` one like the `About PCSX2` dialog has, Orca will continue reading after the user has closed the window.
- [ ] Progress bars give no progress feedback. I'm not sure how continuous progress notification is supposed to work in speech, but there's not even a confirmation that a process like rescanning games or game verification has finished.
- [ ] Sorting the columns in the game list has no feedback (See debugger issue (2)).

## Big Picture Mode

Nothing works. \</section>

## Virtual Machine

- [ ] OSD messages are not read aloud.
- [ ] None of the `Esc` options are read aloud.
- [ ] More of a nice-to-have than something fundamentally broken: things like pausing the emulator just register as the hotkey they're mapped to (Orca announces every key you type) rather than saying something like "Emulator paused". Hotkeys in general do not announce their actions when pressed.

## Debugger

- [ ] Various elements like the disassembly rows are not read out when clicked.
- [ ] Changing the sort on things like thread priority has no feedback (See main window issue (8)).
- [ ] The way the layout locked checkbox changes text leads to the oddity: "Layout Unlocked checkbox not checked" when the layout is unlocked.
- [ ] Clicking the `+` button to create a `New Layout` instead reads the name of the tab the user is currently on (e.g. "R5900 (EE)").

**Comment by TellowKrinkle** [comment]
Here's the state of macOS with VoiceOver:

(Don't forget to unmute the videos, they play muted by default.)

<details>
<summary>First, for reference, here's how a properly working application works</summary>

https://github.com/user-attachments/assets/a88bb931-5b80-4be4-bd26-81c7916091f5
</details>
Things to note:

- The sidebar is a table that you can choose to enter into with VO+Shift+↓.  Once you've entered it, VO+↑↓ selects entries.
- Drop-down menus work as expected.
- Tables read the title and cell when switching left to right, but just the cell contents when switching vertically.  They also read out the size of the table and where you are in it.

<details>
<summary><h3>PCSX2 2.5.258 Settings</h3></summary>

https://github.com/user-attachments/assets/b877bed4-7418-4db5-a4ba-5d9d1b37590c
</details>

- [ ] There's an invisible empty field between Audio and Memory Cards that prevents moving between them with VO+↑↓.  When moving with VO+←→, the cursor goes to the invisible field before moving on to the next item.
- [ ] You can read the different sidebar items but you cannot select them.
- [ ] When attempting to edit drop-down menus, they immediately release focus and you can't edit them.  You can stack multiple on screen at once instead if you want.  Unsurprisingly, no options from the menus are read out either.

<details>
<summary><h3>PCSX2 2.5.258 Main Window</h3></summary>

https://github.com/user-attachments/assets/69448bb2-a23a-4dbb-b3ad-db05ce216383
</details>

- [ ] The multiline-toolbar button and view mode buttons are apparently internally checkboxes, and checking/unchecking them visually changes them in the UI but does not actually invoke their actions
- [ ] When in the game list, VoiceOver always thinks it's in the first column (whose name it doesn't know), regardless of what column it actually is in
- [ ] All the icon-based columns are just read as "empty cell"
- [ ] Some cells are just skipped over for unknown reasons
- [ ] You can't actually interact with any elements in the table.  VO+Shift+M (which is supposed to right click the selected item) does nothing.  VO+Shift+Space (which is supposed to left click it) also does nothing.
- [ ] Selecting table entries with the mouse or PCSX2 keyboard interaction doesn't read anything at all.
